diff -ur xz-5.4.2-orig/build-aux/config.sub xz-5.4.2/build-aux/config.sub
--- xz-5.4.2-orig/build-aux/config.sub	2023-03-19 00:08:55
+++ xz-5.4.2/build-aux/config.sub	2023-04-13 12:54:51
@@ -1126,7 +1126,7 @@
 	xscale-* | xscalee[bl]-*)
 		cpu=`echo "$cpu" | sed 's/^xscale/arm/'`
 		;;
-	arm64-* | aarch64le-*)
+	arm64-* | arm64_32-* | aarch64le-*)
 		cpu=aarch64
 		;;
 
@@ -1728,7 +1728,7 @@
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
-	     | os9* | macos* | osx* | ios* \
+	     | os9* | macos* | osx* | ios* | tvos* | watchos* \
 	     | mpw* | magic* | mmixware* | mon960* | lnews* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
@@ -1791,6 +1791,8 @@
 	os2-emx)
 		;;
 	*-eabi* | *-gnueabi*)
+		;;
+	ios*-simulator | tvos*-simulator | watchos*-simulator)
 		;;
 	-*)
 		# Blank kernel with real OS is always fine.
