--- Python2.7-old/pyconfig.h	2013-08-26 19:26:05.000000000 +0200
+++ Python2.7-new/pyconfig.h	2013-08-27 00:11:06.000000000 +0200
@@ -72,7 +72,7 @@
 /* #undef HAVE_ATTRIBUTE_FORMAT_PARSETUPLE */
 
 /* Define to 1 if you have the `bind_textdomain_codeset' function. */
-#define HAVE_BIND_TEXTDOMAIN_CODESET 1
+/* #undef HAVE_BIND_TEXTDOMAIN_CODESET */
 
 /* Define to 1 if you have the <bluetooth/bluetooth.h> header file. */
 /* #undef HAVE_BLUETOOTH_BLUETOOTH_H */
@@ -218,7 +218,7 @@
 /* #undef HAVE_FDATASYNC */
 
 /* Define to 1 if you have the `finite' function. */
-#define HAVE_FINITE 1
+/* #undef HAVE_FINITE */
 
 /* Define to 1 if you have the `flock' function. */
 #define HAVE_FLOCK 1
@@ -233,7 +233,7 @@
 #define HAVE_FPATHCONF 1
 
 /* Define to 1 if you have the `fseek64' function. */
-#define HAVE_FSEEK64 1
+/* #undef HAVE_FSEEK64 */
 
 /* Define to 1 if you have the `fseeko' function. */
 #define HAVE_FSEEKO 1
@@ -245,7 +245,7 @@
 #define HAVE_FSYNC 1
 
 /* Define to 1 if you have the `ftell64' function. */
-#define HAVE_FTELL64 1
+/* #undef HAVE_FTELL64 */
 
 /* Define to 1 if you have the `ftello' function. */
 #define HAVE_FTELLO 1
@@ -260,7 +260,7 @@
 #define HAVE_GAI_STRERROR 1
 
 /* Define to 1 if you have the `gamma' function. */
-#define HAVE_GAMMA 1
+/* #undef HAVE_GAMMA */
 
 /* Define if we can use gcc inline assembler to get and set x87 control word
    */
@@ -279,10 +279,10 @@
 #define HAVE_GETGROUPS 1
 
 /* Define to 1 if you have the `gethostbyname' function. */
-/* #undef HAVE_GETHOSTBYNAME */
+#define HAVE_GETHOSTBYNAME 1
 
 /* Define this if you have some version of gethostbyname_r() */
-#define HAVE_GETHOSTBYNAME_R 1
+/* #undef HAVE_GETHOSTBYNAME_R */
 
 /* Define this if you have the 3-arg version of gethostbyname_r(). */
 /* #undef HAVE_GETHOSTBYNAME_R_3_ARG */
@@ -291,7 +291,7 @@
 /* #undef HAVE_GETHOSTBYNAME_R_5_ARG */
 
 /* Define this if you have the 6-arg version of gethostbyname_r(). */
-#define HAVE_GETHOSTBYNAME_R_6_ARG 1
+/* #undef HAVE_GETHOSTBYNAME_R_6_ARG */
 
 /* Define to 1 if you have the `getitimer' function. */
 #define HAVE_GETITIMER 1
@@ -327,19 +327,19 @@
 #define HAVE_GETPWENT 1
 
 /* Define to 1 if you have the `getresgid' function. */
-#define HAVE_GETRESGID 1
+/* #undef HAVE_GETRESGID */
 
 /* Define to 1 if you have the `getresuid' function. */
-#define HAVE_GETRESUID 1
+/* #undef HAVE_GETRESUID */
 
 /* Define to 1 if you have the `getsid' function. */
 #define HAVE_GETSID 1
 
 /* Define to 1 if you have the `getspent' function. */
-#define HAVE_GETSPENT 1
+/* #undef HAVE_GETSPENT */
 
 /* Define to 1 if you have the `getspnam' function. */
-#define HAVE_GETSPNAM 1
+/* #undef HAVE_GETSPNAM */
 
 /* Define to 1 if you have the `gettimeofday' function. */
 #define HAVE_GETTIMEOFDAY 1
@@ -465,7 +465,7 @@
 #define HAVE_MKTIME 1
 
 /* Define to 1 if you have the `mremap' function. */
-#define HAVE_MREMAP 1
+/* #undef HAVE_MREMAP */
 
 /* Define to 1 if you have the <ncurses.h> header file. */
 /* #undef HAVE_NCURSES_H */
@@ -492,7 +492,7 @@
 #define HAVE_PAUSE 1
 
 /* Define to 1 if you have the `plock' function. */
-#define HAVE_PLOCK 1
+/* #undef HAVE_PLOCK */
 
 /* Define to 1 if you have the `poll' function. */
 #define HAVE_POLL 1
@@ -567,7 +567,7 @@
 #define HAVE_SEM_OPEN 1
 
 /* Define to 1 if you have the `sem_timedwait' function. */
-#define HAVE_SEM_TIMEDWAIT 1
+/* #undef HAVE_SEM_TIMEDWAIT */
 
 /* Define to 1 if you have the `sem_unlink' function. */
 #define HAVE_SEM_UNLINK 1
@@ -600,10 +600,10 @@
 #define HAVE_SETREGID 1
 
 /* Define to 1 if you have the `setresgid' function. */
-#define HAVE_SETRESGID 1
+/* #undef HAVE_SETRESGID */
 
 /* Define to 1 if you have the `setresuid' function. */
-#define HAVE_SETRESUID 1
+/* #undef HAVE_SETRESUID */
 
 /* Define to 1 if you have the `setreuid' function. */
 #define HAVE_SETREUID 1
@@ -829,7 +829,7 @@
 #define HAVE_TMPNAM 1
 
 /* Define to 1 if you have the `tmpnam_r' function. */
-#define HAVE_TMPNAM_R 1
+/* #undef HAVE_TMPNAM_R */
 
 /* Define to 1 if your `struct tm' has `tm_zone'. Deprecated, use
    `HAVE_STRUCT_TM_TM_ZONE' instead. */
@@ -894,7 +894,7 @@
 #define HAVE_ZLIB_COPY 1
 
 /* Define to 1 if you have the `_getpty' function. */
-#define HAVE__GETPTY 1
+/* #undef HAVE__GETPTY */
 
 /* Define if you are using Mach cthreads directly under /include */
 /* #undef HURD_C_THREADS */
@@ -1245,5 +1245,4 @@
 #define STRICT_SYSV_CURSES /* Don't use ncurses extensions */
 #endif
 
-#endif /*Py_PYCONFIG_H*/
-
+#endif /*Py_PYCONFIG_H*/
\ No newline at end of file
