--- Python-2.7.1/Modules/_sqlite/module.c.orig	2012-10-28 02:30:58.000000000 +0200
+++ Python-2.7.1/Modules/_sqlite/module.c	2012-10-28 02:28:12.000000000 +0200
@@ -28,6 +28,9 @@
 #include "prepare_protocol.h"
 #include "microprotocols.h"
 #include "row.h"
+#ifndef MODULE_NAME
+#define MODULE_NAME "_sqlite3"
+#endif
 
 #if SQLITE_VERSION_NUMBER >= 3003003
 #define HAVE_SHARED_CACHE
--- Python-2.7.1/Modules/_sqlite/sqlitecompat.h.orig	2012-10-28 02:30:53.000000000 +0200
+++ Python-2.7.1/Modules/_sqlite/sqlitecompat.h	2012-10-28 02:28:14.000000000 +0200
@@ -26,6 +26,10 @@
 #ifndef PYSQLITE_COMPAT_H
 #define PYSQLITE_COMPAT_H
 
+#ifndef MODULE_NAME
+#define MODULE_NAME "_sqlite3"
+#endif
+
 /* define Py_ssize_t for pre-2.5 versions of Python */
 
 #if PY_VERSION_HEX < 0x02050000
