diff --git a/Lib/test/test_threading.py b/Lib/test/test_threading.py
index 9c6561c099..f49aafcd9e 100644
--- a/Lib/test/test_threading.py
+++ b/Lib/test/test_threading.py
@@ -1171,6 +1171,8 @@
             os.set_blocking(r, False)
         return (r, w)

+    @unittest.skipIf(sys.platform in ('ios', 'tvos', 'watchos'),
+                     "%s doesn't have os.pipe" % sys.platform)
     def test_threads_join(self):
         # Non-daemon threads should be joined at subinterpreter shutdown
         # (issue #18808)
@@ -1199,6 +1201,8 @@
         # The thread was joined properly.
         self.assertEqual(os.read(r, 1), b"x")

+    @unittest.skipIf(sys.platform in ('ios', 'tvos', 'watchos'),
+                     "%s doesn't have os.pipe" % sys.platform)
     def test_threads_join_2(self):
         # Same as above, but a delay gets introduced after the thread's
         # Python code returned but before the thread state is deleted.
