--- libffi-3.0.13-old/build-ios.sh	2014-10-20 18:42:48.000000000 +0800
+++ libffi-3.0.13-new/build-ios.sh	2014-10-20 18:43:50.000000000 +0800
@@ -2,8 +2,8 @@

 PLATFORM_IOS=/Developer/Platforms/iPhoneOS.platform/
 PLATFORM_IOS_SIM=/Developer/Platforms/iPhoneSimulator.platform/
-SDK_IOS_VERSION="4.2"
-MIN_IOS_VERSION="3.0"
+SDK_IOS_VERSION="6.0"
+MIN_IOS_VERSION="6.0"
 OUTPUT_DIR="universal-ios"

 build_target () {
