--- libffi-3.0.13-old/generate-ios-source-and-headers.py  2014-10-20 21:16:29.000000000 +0800
+++ libffi-3.0.13-new/generate-ios-source-and-headers.py  2014-10-20 21:16:31.000000000 +0800
@@ -125,7 +125,7 @@
         os.makedirs(build_dir)
         env = dict(CC=xcrun_cmd('clang'),
                    LD=xcrun_cmd('ld'),
-                   CFLAGS='-arch %s -isysroot %s -miphoneos-version-min=4.0' % (platform.arch, platform.sdkroot))
+                   CFLAGS='-arch %s -isysroot %s -miphoneos-version-min=6.0' % (platform.arch, platform.sdkroot))
         working_dir=os.getcwd()
         try:
             os.chdir(build_dir)
