--- libffi-3.1-old/generate-darwin-source-and-headers.py  2014-10-23 17:49:33.000000000 +0800
+++ libffi-3.1-new/generate-darwin-source-and-headers.py  2014-10-23 17:50:01.000000000 +0800
@@ -14,7 +14,7 @@
     sdk = 'iphonesimulator'
     arch = 'i386'
     triple = 'i386-apple-darwin11'
-    version_min = '-miphoneos-version-min=5.1.1'
+    version_min = '-miphoneos-version-min=6.0'

     prefix = "#ifdef __i386__\n\n"
     suffix = "\n\n#endif"
@@ -40,7 +40,7 @@
     sdk = 'iphoneos'
     arch = 'armv7'
     triple = 'arm-apple-darwin11'
-    version_min = '-miphoneos-version-min=5.1.1'
+    version_min = '-miphoneos-version-min=6.0'

     prefix = "#ifdef __arm__\n\n"
     suffix = "\n\n#endif"
