diff --git a/generate-darwin-source-and-headers.py b/generate-darwin-source-and-headers.py
index f7fc414..516464f 100755
--- a/generate-darwin-source-and-headers.py
+++ b/generate-darwin-source-and-headers.py
@@ -143,7 +143,7 @@ def build_target(platform, platform_headers):
     mkdir_p(build_dir)
     env = dict(CC=xcrun_cmd('clang'),
                LD=xcrun_cmd('ld'),
-               CFLAGS='%s' % (platform.version_min))
+               CFLAGS='%s -fembed-bitcode' % (platform.version_min))
     working_dir = os.getcwd()
     try:
         os.chdir(build_dir)
@@ -173,7 +173,6 @@ def generate_source_and_headers(generate_osx=True, generate_ios=True):
         copy_src_platform_files(device_platform)
         copy_src_platform_files(device64_platform)
     if generate_osx:
-        copy_src_platform_files(desktop32_platform)
         copy_src_platform_files(desktop64_platform)
 
     platform_headers = collections.defaultdict(set)
@@ -184,11 +183,10 @@ def generate_source_and_headers(generate_osx=True, generate_ios=True):
         build_target(device_platform, platform_headers)
         build_target(device64_platform, platform_headers)
     if generate_osx:
-        build_target(desktop32_platform, platform_headers)
         build_target(desktop64_platform, platform_headers)
 
     mkdir_p('darwin_common/include')
-    for header_name, tag_tuples in platform_headers.iteritems():
+    for header_name, tag_tuples in platform_headers.items():
         basename, suffix = os.path.splitext(header_name)
         with open(os.path.join('darwin_common/include', header_name), 'w') as header:
             for tag_tuple in tag_tuples:
